### ISSUE TITLE
setRecordVersionAndTimestamp is not called when using batch insert or update

### ISSUE THREAD

**Issue by SButterfly** [jOOQ/jOOQ#11937]
I am using recordTimestampFields. And when I insert or update single record new timestamp is set to record after update.
But when batch insert/update is used changed data not saved to records.

### Expected behavior
timestamp is saved to record in batch insert/updated with recordTimestampFields enabled

### Actual behavior
it is not saved

### Steps to reproduce the problem
1. Create several records with recordTimestampFields is one
2. Call `dslContext.batchInsert(records).execute()`

### Versions

- jOOQ: 3.13.2
- Java: 11
- Database (include vendor): postgres
- OS: macos
- JDBC Driver (include name if inofficial driver): postgresql-42.2.8.fix1

**Comment by lukaseder** [comment]
Thanks a lot for your report. We'll look into this soon. In the meantime, you might get better results by using the batched connection, which batches things on the JDBC level, not on the jOOQ API level:
https://www.jooq.org/doc/latest/manual/sql-execution/batched-connection/

**Comment by SButterfly** [comment]
I've tried to change code, but still no expected result
```
            List<? extends Query> inserts = records.stream()
                .map(r -> dslContext.insertInto(table).set(r))
                .collect(Collectors.toList());
            dslContext.batch(inserts).execute();
```

**Comment by lukaseder** [comment]
> I've tried to change code, but still no expected result

That's not what that page is suggesting. It suggests you run:

```java
dslContext.batched(c -> {
    records.forEach(r -> {
        r.insert();
    });
});
```

**Comment by SButterfly** [comment]
Thanks, but it doesn't work.
Firstly, it does't run my inserts as batch.
```
2021-06-01 17:50:44,146 DEBUG [LoggerListener main] Executing query          : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 17:50:44,148 DEBUG [LoggerListener main] -> with bind values      : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (1, 'A', timestamp '2021-06-01 17:50:44.128', null, null, null, null) returning "test"."jooq_test_entity"."id"
2021-06-01 17:50:44,165 DEBUG [LoggerListener main] Fetched result           : +----+
2021-06-01 17:50:44,165 DEBUG [LoggerListener main]                          : |  id|
2021-06-01 17:50:44,165 DEBUG [LoggerListener main]                          : +----+
2021-06-01 17:50:44,165 DEBUG [LoggerListener main]                          : |   1|
2021-06-01 17:50:44,165 DEBUG [LoggerListener main]                          : +----+
2021-06-01 17:50:44,166 DEBUG [LoggerListener main] Fetched row(s)           : 1
2021-06-01 17:50:44,171 DEBUG [LoggerListener main] Executing query          : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 17:50:44,172 DEBUG [LoggerListener main] -> with bind values      : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (2, 'A', timestamp '2021-06-01 17:50:44.168', null, null, null, null) returning "test"."jooq_test_entity"."id"
2021-06-01 17:50:44,173 DEBUG [LoggerListener main] Fetched result           : +----+
2021-06-01 17:50:44,173 DEBUG [LoggerListener main]                          : |  id|
2021-06-01 17:50:44,173 DEBUG [LoggerListener main]                          : +----+
2021-06-01 17:50:44,173 DEBUG [LoggerListener main]                          : |   2|
2021-06-01 17:50:44,174 DEBUG [LoggerListener main]                          : +----+
2021-06-01 17:50:44,174 DEBUG [LoggerListener main] Fetched row(s)           : 1
```

Secondly, how can I get number of changed rows? Previously I got them like this:
```
int[] result = dslContext.batchInsert(records).execute();
```

**Comment by lukaseder** [comment]
> Firstly, it does't run my inserts as batch.

Are you sure? Indeed, we should improve the logging, but there should be a batch executed: https://github.com/jOOQ/jOOQ/issues/11467

Notice I attached each record to the new "batched" `Configuration`

> Secondly, how can I get number of changed rows? Previously I got them like this:

You can count the records in this case.

Anyway. This is just a workaround. We'll definitely fix the actual issue eventually.

**Comment by SButterfly** [comment]
First time i missed line with attach function.
I again tried the code you published:
```java
dslContext.batched(c -> {
                records.forEach(r -> {
                    r.attach(c);
                    r.insert();
                });
            });
```
And it still doesn't work in batch.
```
2021-06-01 18:31:06,911 DEBUG [LoggerListener main] Executing query          : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,911 DEBUG [LoggerListener main] -> with bind values      : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (1, 'A', timestamp '2021-06-01 18:31:06.905', null, null, null, null) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,918 DEBUG [BatchedPreparedStatement main] BatchedStatement         : Skipped batching statement: insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,921 DEBUG [LoggerListener main] Fetched result           : +----+
2021-06-01 18:31:06,921 DEBUG [LoggerListener main]                          : |  id|
2021-06-01 18:31:06,921 DEBUG [LoggerListener main]                          : +----+
2021-06-01 18:31:06,921 DEBUG [LoggerListener main]                          : |   1|
2021-06-01 18:31:06,921 DEBUG [LoggerListener main]                          : +----+
2021-06-01 18:31:06,921 DEBUG [LoggerListener main] Fetched row(s)           : 1
2021-06-01 18:31:06,923 DEBUG [LoggerListener main] Executing query          : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,923 DEBUG [LoggerListener main] -> with bind values      : insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (2, 'A', timestamp '2021-06-01 18:31:06.922', null, null, null, null) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,923 DEBUG [BatchedPreparedStatement main] BatchedStatement         : Skipped batching statement: insert into "test"."jooq_test_entity" ("id", "description", "modified_date", "from_date", "to_date", "created_at", "created_at_tz") values (?, ?, cast(? as timestamp(6)), cast(? as date), cast(? as date), cast(? as timestamp(6)), cast(? as timestamp(6) with time zone)) returning "test"."jooq_test_entity"."id"
2021-06-01 18:31:06,924 DEBUG [LoggerListener main] Fetched result           : +----+
2021-06-01 18:31:06,924 DEBUG [LoggerListener main]                          : |  id|
2021-06-01 18:31:06,924 DEBUG [LoggerListener main]                          : +----+
2021-06-01 18:31:06,924 DEBUG [LoggerListener main]                          : |   2|
2021-06-01 18:31:06,924 DEBUG [LoggerListener main]                          : +----+
2021-06-01 18:31:06,924 DEBUG [LoggerListener main] Fetched row(s)           : 1
```

I saw **Skipped batching statement:**. 
After some time of research I found out, that this line of code produce non-batch insert https://github.com/jOOQ/jOOQ/blob/e426c10fbc618b6cef77d5879f72ff73121a07ff/jOOQ/src/main/java/org/jooq/impl/AbstractDMLQuery.java#L1041

As far as I understand, it is due `returning` part in my sql query. But comment is not quite true. Actually, you _can_ get result of generated values in batch insert.
Here is very simple code how you can do it in spring-jdbc
```java
       return JdbcOperations.execute((ConnectionCallback<int[]>) con -> {
            try (PreparedStatement statement = con.prepareStatement("insert into test.jooq_test_entity (description) values (?) returning id", new String[]{ "id" })) {
                for (var record : records) {
                    int i = 0;
                    statement.setString(++i, "description");
                    statement.addBatch();
                }
                int[] updates = statement.executeBatch();

                ResultSet resultSet = statement.getGeneratedKeys();
                long newId = resultSet.getLong(0);
                // got auto-generated values
                return updates;
            }
        });
```

-------

Here is my table
```sql
create table test.jooq_test_entity (
    id          bigserial primary key,
    description text,
    modified_date timestamp
);
alter table test.jooq_test_entity add from_date date;
alter table test.jooq_test_entity add to_date date;
alter table test.jooq_test_entity add created_at timestamp;
alter table test.jooq_test_entity add created_at_tz timestamp with time zone;
```
And [jooq.txt](https://github.com/jOOQ/jOOQ/files/6577235/jooq.txt)

**Comment by SButterfly** [comment]
> You can count the records in this case.

I don't understand how is it even possible in such code. What if I have insert/update with `update ... where modif_ts >= ?` condition?
```java
            dslContext.batched(c -> {
                records.forEach(r -> {
                    r.attach(c);
                    int result = r.insert(); // what will be result in batch insert?
                    count += result;
                });
            });
            if (count != records.size()) {
                throw new NotAllInsertedException();
            }
```

**Comment by lukaseder** [comment]
Related https://github.com/jOOQ/jOOQ/issues/8283

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/8283 has been fixed for jOOQ versions 3.19.0, 3.18.8, 3.17.17, 3.16.23. I think that this is a duplicate of #8283.
